File: test-vine-task-throughput.c
#include "taskvine.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <time.h>
#include <assert.h>
int main(int argc, char *argv[])
{
	struct vine_manager *m;
	int i;
    int tasksC = 5000;

	m = vine_create(VINE_DEFAULT_PORT);

	if(!m) {
		printf("couldn't create manager: %s\n", strerror(errno));
		return 1;
	}
	printf("TaskVine listening on %d\n", vine_port(m));

	for(i=0;i<tasksC;i++) {
		struct vine_task *t = vine_task_create(":");
		vine_task_set_cores(t, 1);

		int task_id = vine_submit(m, t);

	}
	bool start_timer = true;
	clock_t start = 0;
	while(!vine_empty(m)) {
        struct vine_task *t  = vine_wait(m, 5);
		if (start_timer){
			start = clock();
			start_timer = false;
		}
        if(t) {
            vine_task_delete(t);
        }
    }
	clock_t end = clock();
	double throughtput_time = ((double)(end - start)) / CLOCKS_PER_SEC;

	double throughput = tasksC / throughtput_time;
	
	start_timer = true;
	start = 0;
	for(i=0;i<tasksC;i++) {
		struct vine_task *t = vine_task_create(":");
		vine_task_set_cores(t, 1);
		if (start_timer){
			start = clock();
			start_timer = false;
		}
		int task_id = vine_submit(m, t);
        while(!vine_empty(m)) {
            t  = vine_wait(m, 5);
            if(t) {
                vine_task_delete(t);
            }
        }
	}
	end = clock();
	double chaining_time = ((double)(end - start)) / CLOCKS_PER_SEC;
	double chaining = tasksC / chaining_time;

	printf("Throughput was %f tasks per second\n", throughput);
	assert(throughput > 900);

	printf("Chaining was %f tasks per second\n", chaining);
	printf("all tasks complete!\n");
	assert(chaining > 1500);
	vine_delete(m);

	return 0;
}

File: test-vine-task-throughput.py
#!/usr/bin/env python

import ndcctools.taskvine as vine
import time


def func():
    return


def main():

    q = vine.Manager()
    print("Creating Library from functions...")
    function_lib = q.create_library_from_functions("test-library", func, add_env=False)
    q.install_library(function_lib)

    print("listening on port", q.port)
    factory = vine.Factory("local", manager_host_port="localhost:{}".format(q.port))
    factory.max_workers = 1
    factory.min_workers = 1

    num_tasks = 1000

    with factory:
        for i in range(num_tasks):
            t = vine.Task(command=":")
            task_id = q.submit(t)

        print("waiting for tasks to complete...")

        start_timer = True
        while not q.empty():

            t = q.wait(5)
            if start_timer:
                start = time.time()
                start_timer = False

        end = time.time()
        many = end - start

        start = time.time()

        for i in range(num_tasks):
            while not q.empty():
                result = q.wait(5)
            t = vine.Task(command=":")
            task_id = q.submit(t)

        print("waiting for tasks to complete...")
        end = time.time()
        one = end - start
        throughput = num_tasks / many
        chaining = num_tasks / one
        # serverless tasks
        for i in range(num_tasks):
            t = vine.FunctionCall("test-library", "func")
            task_id = q.submit(t)

        print("Waiting for tasks to complete...")
        start_timer = True
        while not q.empty():
            t = q.wait(5)
            if start_timer:
                start = time.time()
                start_timer = False
        end = time.time()
        serverless_many = end - start

        start = time.time()
        for i in range(num_tasks):
            while not q.empty():
                q.wait(5)
            t = vine.FunctionCall("test-library", "func")
            q.submit(t)
        end = time.time()
        serverless_one = end - start
    serverless_throughput = num_tasks / serverless_many
    serverless_chaining = num_tasks / serverless_one

    print(f"\nThroughput was {throughput} tasks per second")
    print(f"Chaining was {chaining} tasks per second")
    print(f"Serverless Throughput was {serverless_throughput} tasks per second")
    print(f"Serverless Chaining was {serverless_chaining} tasks per second")
    print("all tasks complete!")


if __name__ == "__main__":
    main()
